cmd/snapd/main.go

```diff
@@ -52,6 +54,36 @@ func main() {
 	}
 }
 
+func runWatchdog(d *daemon.Daemon) (*time.Ticker, error) {
+	// not running under systemd
+	if os.Getenv("WATCHDOG_USEC") == "" {
+		return nil, nil
+	}
+	usec := osutil.GetenvInt64("WATCHDOG_USEC")
+	if usec == 0 {
+		return nil, fmt.Errorf("cannot parse WATCHDOG_USEC: %q", os.Getenv("WATCHDOG_USEC"))
+	}
+	dur := time.Duration(usec/2) * time.Microsecond
+	logger.Debugf("Setting up sd_notify() watchdog timer every %s", dur)
+	wt := time.NewTicker(dur)
+
+	go func() {
+		for {
+			select {
+			case <-wt.C:
+				// TODO: poke the snapd API here and
+				//       only report WATCHDOG=1 if it
+				//       replies with valid data
+				systemd.SdNotify("WATCHDOG=1")
```

Comment:
Another potential TODO (also low-tech and cheap), add a counter to the Ensure loop in overlord increment it with each iteration and monitor from there it's changing.